### ISSUE TITLE
Writting dependent Specs?

### ISSUE THREAD

**Issue by Julio-Guerra** [onsi/ginkgo#246]
[Given the example from the documentation](https://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context), we can see an example of dependent specs: if the call to `NewBookFromJSON()` fails, everything else will too, and with a panic. This is a very common case of dependent specs and we can of course imagine others.

As far as I know, these dependencies make impossible running a suite in parallel or randomized order. I firstly thought the Specs were randomized/run within **a single container only**, but it appears it is global. 

Here is my personal example in pseudo-code which fails in a parallel run:

```
Describe: object creation
    Context: without errors
        It: should not fail
        It: should not be nil

    Describe: given a successfully created objet
        Context: method A
            It: should not fail
            ....
```

Running this is with `randomizeAllSpecs` can show the call to `method A` performed before the specs checking its correct creation.

Do you confirm that this does not express a dependency between the inner Decribe()/Context() and the outer - as I thought? And if not, what is the official position on this?

**Comment by onsi** [comment]
The individual `It`s are required to be independent to make parallelization and randomization possible.  While you can inadvertently affect one test by leaking global state from another test this is almost always a bug in how the tests are written.

The correct approach, in Ginkgo, to do what you are trying to do here is to use `BeforeEach`.  `BeforeEach`es in parent Describe/Context blocks always run Before Each It.

In the docs example, if NewBookFromJSON returns N error, then a common pattern is to add:
`Expect(err).NotTo(HaveOccured())`

To the BeforeEach.  This would correctly fail any tests that rely on the successful creation of the object.

Does that help?

**Comment by Julio-Guerra** [comment]
Yes. This is how I also solved the problem for object creations.
But as soon as you allow the nesting of spec-containers, you kind of express a dependency relationship that `It()` does not really follow because they are kind of "flattened".

Maybe this could be listed as a limitation?

Anyway, thanks for "the official position" :+1:

**Comment by Julio-Guerra** [comment]
I have now written more advanced integration tests using ginkgo and I am more and more convinced that ginkgo should implement nested requirements to be able to write nice readable scenarios which are basically cascaded requirements.

Here is a concrete example of how I mapped into ginkgo's dsl a subset of the TTY API which is full of nested requirements (A => B => C ...): 

```
Linux TTY API
├── Describe: opening a tty device
├──── It: should successfully open the TTY
├──── Describe: reading
├────── Context: a non-empty tty receive buffer
├──────── Context: using the blocking mode
├────────── Context: given a tty receive buffer below the blocking threshold
├──────────── It: should block
├──────────── Context: given a tty receive buffer above the blocking threshold
├────────────── Context: into a user buffer of any size
├────────────── It: should unblock
├────────────── It: should not fail
├────────────── It: should copy the receive up to the user buffer's size
```

It is already possible to get the behavior I am willing by not enabling randomization, nor parallelization and by using the `-failFast` option. That way, the order is preserved and the ginkgo stops as soon as any spec fails. But this is an implementation defined behavior...

@onsi What do you think?

**Comment by robdimsdale** [comment]
As you've seen, this works. I doubt it will change, but it's probably also not part of the official spec (i.e. if there's a really good reason to randomize we might start doing that by default).

I would really encourage you to not couple your tests, as you will never be able to run them randomly or in parallel, two behaviors which we find incredible valuable in writing effective tests and driving out race conditions.

A nice heuristic I've found is that each `Context` should also have a `BeforeEach` otherwise it's not really a new context (you're not modifying the environment from the previous context).

To take your example above, I would do something like:

```
Linux TTY API
Describe: opening a tty device
  It: should successfully open the TTY

  Context: when opening fails
    It: handles the error

  Describe: reading
    BeforeEach: open the TTY, validate no error # similar to the successful It above

    Context: a non-empty tty receive buffer
      Context: using the blocking mode
        Context: given a tty receive buffer below the blocking threshold
          It: should block

        Context: given a tty receive buffer above the blocking threshold
          BeforeEach: setup a receive buffer above the blocking threshold

          Context: into a user buffer of any size
            BeforeEach: set a user buffer of any size
            It: should copy the receive up to the user buffer's size without error #unblocking comes for free
```

I can see the attractiveness in using the fact that one test passed as a way to setup context for another, but in our experiences, when we've written tests this way they become much more prone to race conditions and incorrect test setup (plus without parallelization they run slower).

I accept that we may need to agree to disagree on this.

**Comment by Julio-Guerra** [comment]
I agree and what I am trying to show is that a parent `It()` should have precedence over its siblings containers: recursively, do not run until sibling `It()`s have successfully finished.

It does not prevent doing parallelization nor randomization at all. And it avoids copying/pasting parent `It()`s assertions into sibling and sub-containers' `BeforeEach()`es.

**Comment by robdimsdale** [comment]
By preventing some tests from running until another test is complete, you are preventing parallelization and, to some degree, randomization.

I agree that it avoids copying any pasting, but it does so at the cost of making the tests more brittle, and, to many, more confusing. I accept to you this may be less confusing.

Like I said, we may need to agree to disagree, but it's unlikely that we will modify ginkgo to support this behavior. If you'd like to submit a PR to the docs to clarify this behavior we can look at that.

**Comment by onsi** [comment]
Closing this out - I think @robdimsdale has covered the philosophy and thinking behind this well.
